### ISSUE TITLE
css is static with hard-coded cdn urls

### ISSUE THREAD

**Issue by adamrneary** [activecell/activecell.com#6]
The existing liquid filter only applies to templates and not assets, so the css files still point to local copies.

We either need to hard-code the cdn into all the asset urls or we need to generate the css dynamically. Since we're already using bootstrap, we're sort of already generating css dynamically, but it's not built into the ongoing deploy process.

When it's time to take up this issue, let's decide which approach to take and how to get it done without a lot of hassle, ideally using tools already in place...

Potential plug-ins:
- https://github.com/zroger/jekyll-less#readme
- https://gist.github.com/960150
- others at https://github.com/mojombo/jekyll/wiki/Plugins

**Comment by adamrneary** [comment]
https://gist.github.com/960150 seems like the right above given that we're using Sass for the core product itself. When we get to this, we should just add the Sass plugin and then abstract out the hard-coded CDN URL components. For now, they're just baked right into every asset url.

**Comment by umarsheikh** [comment]
I read the question and answer on http://stackoverflow.com/questions/3848147/image-paths-in-css-to-support-cdn , if my interpretation of the answer is correct, then we are already serving the css file from the cdn. In that case, the image paths and urls in the css files are pointing to "local copies" on the cdn, which means, this problem is already solved. This is the quote from the answer:
"Probably the easiest thing would be to host both your images and CSS file on the CDN. Image paths in your CSS file are relative to the CSS file itself, so you won't have to change your CSS at all."
I will explore this issue for a while and see if I am correct in believing that relative paths in the css file are pointing to images on the cdn itself and not to the ones on the s3 amazon bucket.

**Comment by umarsheikh** [comment]
in order to check this, we need to identify an element on the site, which has a background image. its url would be something like http://d2h0dyinjrpfzl.cloudfront.net/assets/img/bg_quartertwirl.png , then we only need to change it to /assets/img/bg_quartertwirl.png and try and open it. if its url is http://d2h0dyinjrpfzl.cloudfront.net/assets/img/bg_quartertwirl.png then it means that our hunch is correct and images are already being served from the cdn. the problem is there are tons on elements in the css files with background images and most of them are non-existant, so there is lots of redundant css, so it is hard to come up with an element with a background image. i am still looking...

**Comment by umarsheikh** [comment]
according to http://stackoverflow.com/questions/940451/using-relative-url-in-css-file-what-location-is-it-relative-to the image paths etc are relative to the stylesheet, and since the stylesheet is already on the cdn, our assets are being served from the cdn.

**Comment by adamrneary** [comment]
If this is the case we should be able to update the urls in profitably.css accordingly, deploy, and see everything just fine. Could you make those changes and test?

**Comment by umarsheikh** [comment]
sure, one question: we used to deploy with 'jekyll-s3' before, and now we do 'rake deploy'. originally the cdn links weren't working. have u now set them up so that if we now deploy to s3, then assets are automatically deployed to the cdn? i am assuming that for that to happen u would have logged in on s3 with you account and set up some stuff there?

**Comment by umarsheikh** [comment]
I am getting stuck with what to put as the bucket name, and how to use that bucket. For instance, when we were using jekyll-s3, our bucket name was "www.profitably.com.s3-website-us-east-1.amazonaws.com" and it used to deploy to http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/index.html. So i put "s3://www.profitably.com.s3-website-us-east-1.amazonaws.com" as the bucket name as per instructions, but i am not sure where it is deploying when we are using s3cmd. Further, I am not sure if doing a simple deploy also updates the cdn, or you did it manually? so we need to figure this step out. I can push the changes to the deploy_update branch, and you can deploy and see if it works correctly or not.

**Comment by umarsheikh** [comment]
can u create a restricted account for me on amazon so i can log in and see what our bucket endpoints are, cdn information etc?

**Comment by umarsheikh** [comment]
can u create a restricted account for me on amazon so i can log in and see what our bucket endpoints are, cdn information etc?

**Comment by adamrneary** [comment]
you don't have to worry about cloudfront (cdn). once something is uploaded to s3 it is mirrored on the cdn automatically.

**Comment by adamrneary** [comment]
@umar right now when I run rake deploy it's updating the files on s3 but it putting in local urls into the css. It actually has my directories in the files. Can you look into this? Everything looks fine on my machine, but I suspect it's wrong everywhere else!

**Comment by adamrneary** [comment]
@umar also--it doesn't seem to actually be uploading packaged_css_20120515.css. not sure why. if you're around we can chat over hipchat to get this resolved...

**Comment by umarsheikh** [comment]
@adamrneary I simply added jekyll-s3 back, and did a deploy with jekyll-s3. The s3 bucket was www.profitably.com.s3-website-us-east-1.amazonaws.com and the site runs at http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/index.html. but it is unstyled since it is trying to point to http://d2h0dyinjrpfzl.cloudfront.net/assets/packaged_css_20120515.css but that file is not there on the cdn. however, this file is available at http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/assets/packaged_css_20120516.css so even though the file is on s3, it is not automatically on the cdn.

I am now goign to change the packaged file names to 20120516 and deploy with rake deploy. the bucket name stays the same, but i am not sure where it will get deployed to, that is, the endpoint. i am assuming it is the same as above endpoint. if we see the file referenced packaged_css_20120516.css on the endpoint, we will know that the deploy is working but not the cdn. i will be doing now and checking...

**Comment by umarsheikh** [comment]
Once i run rake deploy, i am assuming that the site will get deployed to http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/index.html as above, however, once i run rake deploy, i get this output above:
    < Error>< Code>AccessDenied< /Code>< Message>Access Denied< /Message>  
    < RequestId>BDF0027B3E689E2E< /RequestId>  
    < HostId>shP/IvCVniPNk6JX2EJ8Er7T9yAA/hYIapYQHkEygmAH1rCzlVxOyUAYliOwD8HW< /HostId>< /Error>

this means that the site at least was deployed, however there could be a permissions issue. why is it deployed? because after running jekyll-s3 it was at least showing an unstyled page, but after running rake deploy the same page gives an error message. so this looks like a permissions issue to me.

**Comment by umarsheikh** [comment]
Yep, it is a permissions issue. There is a setacl command to change permissions. for testing purposes, i used it as follows:
 s3cmd setacl --recursive —acl-public s3://www.profitably.com.s3-website-us-east-1.amazonaws.com/

once that is done, i see the site deployed as before, and it is trying to reference http://d2h0dyinjrpfzl.cloudfront.net/assets/packaged_css_20120516.css as it should. but this file is not on the cdn. however, if u try to access this file as http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/assets/packaged_css_20120516.css then you see the file. so the deploy is working fine, we need to do the following:
- resolve permissions issues
- make sure the files are also on the cdn.

**Comment by umarsheikh** [comment]
similarly, this is accessible: http://d2h0dyinjrpfzl.cloudfront.net/assets/packaged_css_20120507.css but this is not:
http://d2h0dyinjrpfzl.cloudfront.net/assets/packaged_css_20120516.css as it results in AccessDenied xml response.

**Comment by umarsheikh** [comment]
so summary of findings so far:
- wehn we run jekyll-s3, the site is deployed to http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/index.html
- the site is unstyled, since the css files on the cdn are not working, and giving a permissions issue.

And when we deploy with rake deploy:
- the site is deployed to same location, that is the site is deployed to http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com/index.html
- however, nothing works because of some wrong permissions. to set these permissions up, i simply ran s3cmd setacl --recursive --acl-public s3://www.profitably.com.s3-website-us-east-1.amazonaws.com/ but we need to figure out what the correct permissions should be in order that everyone else can just have read permissions
- after changing permisssions, the site is there, but again, the cdn links to assets are resulting in permission denied.

so it looks like both jekyll-s3 and rake deploy (with the setacl command) are deploying, but the cdn links are not working for now.

**Comment by adamrneary** [comment]
@umarsheikh sorry for the delay on this. We've been crazy busy.

The problem seemed be the bucket name. In the S3 world, the bucket name is "www.profitably.com.s3.amazonaws.com" and then they have a separate endpoint for actually accessing the site as if it were a standalone static site, and that's the "http://www.profitably.com.s3-website-us-east-1.amazonaws.com.s3.amazonaws.com" or whatever bucket. You cannot sync with the website address--only the source bucket. Cloudfront points to the source bucket, so that explains why you weren't seeing files you synced on the cloudfront CDN. I updated the rakefile and it seems to be working...

**Comment by adamrneary** [comment]
@umarsheikh I think we actually need to solve issue #14 before we can solve this one, since you may well be correct that it honors relative references, even on the CDN. Have you taken a look at #14?

**Comment by adamrneary** [comment]
@umarsheikh I made a ton of progress today (or so it feels!). We have production live right on S3 with nothing pointing to github pages. I used dnsimple because they have a url redirect for the apex domain, and that sends people from profitably.com to www.profitably.com. Done and done. Issue #14 closed.

Now we're back to a tricky problem that is the original problem for this bug. During rake:deploy, Jammit is rewriting the css paths as local paths (see the comment chain above). It sounds like we'll be ok with relative asset paths, but Jammit is rewriting them.

http://code.dblock.org/rails-s3-cloudfront-jammit-heroku-100

The above is a great post on this, and dblock seems to have his update merged into jammit's master, but even running his for locally, i was getting the same problem.

Any thoughts?

**Comment by umarsheikh** [comment]
It worked for me! I updated the Gemfile as above, and i ran the command with bundle exec: bundle exec jekyll --server and i got the packaged_css files referencing assets like follows: background-image:url(../img/glyphicons-halflings.png) instead of: 
background-image:url(../home/umar/projects/profitably/secondtask/profitably.github.com/_raw/img/glyphicons-halflings.png)

I also noticed that after installing the jammit gem with the git option, jekyll --server would still generate with the expanded local path, so i had to use bundle exec jekyll --server to fix it.

This case now looks resolved!
